### PR TITLE
[BEAM-7039] set up validatesPortableRunner tests for Spark

### PR DESCRIPTION
**Pull request by ibzib**
This adds a couple useful Gradle commands for the Spark portable runner:
1. `:beam-runners-spark-job-server:runShadow`
2. `:beam-runners-spark-job-server:validatesPortableRunner`

At this commit, most tests fail for the latter because PAssert is not yet supported (pending Metrics & Flatten).

Disclaimer: I am not too familiar with Gradle/Groovy, so please make extra sure I didn't make any rookie mistakes :)

R: @mxm 

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by ibzib** [comment]
Run Portable_Python PreCommit

**Comment by ibzib** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by ibzib** [comment]
Run Java PreCommit

File: runners/spark/job-server/build.gradle [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+// task will not work because the Spark runner classes only exist in the shadow
+// jar.
+runShadow {
+  args = []
+  if (project.hasProperty('jobHost'))
+    args += ["--job-host=${project.property('jobHost')}"]
+  if (project.hasProperty('artifactsDir'))
+    args += ["--artifacts-dir=${project.property('artifactsDir')}"]
+  if (project.hasProperty('cleanArtifactsPerJob'))
+    args += ["--clean-artifacts-per-job"]
```
Comment (mxm) [inline runners/spark/job-server/build.gradle]: This is actually outdated. As of #8210, this is always true. We should parameterize the option to allow it to be set to false/true. I think this also needs to be fixed for Flink.

File: runners/spark/job-server/build.gradle [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  jvmArgs = ["-Xdebug", "-Xrunjdwp:transport=dt_socket,server=y,suspend=n,address=5005"]
+  if (project.hasProperty("logLevel"))
+    jvmArgs += ["-Dorg.slf4j.simpleLogger.defaultLogLevel=${project.property('logLevel')}"]
+}
+
+def portableValidatesRunnerTask(String name) {
+  createPortableValidatesRunnerTask(
+    name: "validatesPortableRunner${name}",
+    jobServerDriver: "org.apache.beam.runners.spark.SparkJobServerDriver",
+    jobServerConfig: "--clean-artifacts-per-job,--job-host=localhost,--job-port=0,--artifact-port=0",
```
Comment (mxm) [inline runners/spark/job-server/build.gradle]: We can remove the clean artifacts option here.

File: runners/spark/job-server/build.gradle (hunk 1 above)
Comment (ibzib) [reply runners/spark/job-server/build.gradle]: fixed

Comment (ibzib) [reply runners/spark/job-server/build.gradle]: I made a PR to fix this for Flink as well. #8293

**Comment by ibzib** [comment]
Run Website PreCommit

**Comment by ibzib** [comment]
Run Python PreCommit

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 3]
```diff
@@ -1559,6 +1561,7 @@ class BeamModulePlugin implements Plugin<Project> {
       project.tasks.create(name: name, type: Test) {
         group = "Verification"
         description = "Validates the PortableRunner with JobServer ${config.jobServerDriver}"
+        systemProperties config.systemProperties
         systemProperty "beamTestPipelineOptions", JsonOutput.toJson(beamTestPipelineOptions)
```
Comment (angoenka) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Shall we also merge this system property to systemProperties set above,

File: runners/spark/job-server/build.gradle (hunk 1 above)
Comment (mxm) [reply runners/spark/job-server/build.gradle]: Thanks for following up with this!

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 3 above)
Comment (ibzib) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: updated

**Comment by ibzib** [comment]
Run Java PreCommit

**Comment by ibzib** [comment]
Run Website PreCommit

**Comment by ibzib** [comment]
Run Website PreCommit

**Comment by mxm** [comment]
Needs to be rebased now due to a merge conflict.

**Comment by ibzib** [comment]
> Needs to be rebased now due to a merge conflict.

resolved

**Comment by ibzib** [comment]
Run Portable_Python PreCommit

**Comment by mxm** [comment]
Thanks!

Website test failures unrelated.

**Pull request merged**